### DISCUSSION TITLE
Setting Request Headers with Request

### DISCUSSION THREAD

**Discussion by shellscape** [nodejs/undici#1168] [Q&A]
## This would solve...

Since `fetch` is only available on Node v16.something+, we're unable to use `fetch` on AWS Lambda (without jumping through very slow hoops) and systems that are locked to Node v14.x.

## The implementation should look like...

Not sure, as I'm unfamiliar with the internals. But expanding the `options` interface seems appropriate.

## I have also considered...

Not using undici, but that'd be a real bummer.

## Additional context

If there's an alternative method to setting headers with `request`, it's not in the documentation, and would be great to have added. I'm a long-time `got` user, but I'm compelled by what undici is accomplishing.

**Comment by shellscape** [comment]
@ronag this isn't a discussion, it's a feature request. the signature of `request` does not allow passing `headers` as an option. Why was this converted?

**Reply by ronag** [reply]
The docs does state you can pass a headers options. https://undici.nodejs.org/#/docs/api/Dispatcher?id=parameter-dispatchoptions

**Reply by shellscape** [reply]
This is a documentation issue then (ultimately because I guess I didn't take my time or missed a some small text), resulting from some documentation and typing convolution. Again, converting this to a discussion was premature, and a bummer.

https://undici.nodejs.org/#/?id=undicirequesturl-options-promise status that `options RequestOptions` and links to https://undici.nodejs.org/#/docs/api/Dispatcher?id=parameter-requestoptions for which there's a _teensy little ditty_ about `Extends: DispatchOptions`. That could be made _much_ more visible and verbose. e.g. `The options object may also contain options from DispatchOptions` so that it doesn't go missed. Documentation doesn't have be terse and can be a lovely experience to read.

There's also the matter of the typings. The request typings read:

```
options?: { dispatcher?: Dispatcher } & Omit<Dispatcher.RequestOptions, 'origin' | 'path'>,
```

That's an intersection, not an extension. So the documentation isn't correct that the options for `request` are `RequestOptions`. Semantic hubalaboo but it's still not correct. It can however, lead to confusion when reading the typings within an editor. 

With the typings there's some inconsistency between `fetch` options and `request` options: `request` options require passing the `method: HttpMethod` property, `fetch` doesn't require this.

**Reply by ronag** [reply]
Yes, good suggestions. PR welcome!

**Reply by shellscape** [reply]
Slide this back into issues and I'll be happy to work on a contribution.

**Reply by ronag** [reply]
https://github.com/nodejs/undici/issues/1169
